### PR TITLE
Fix Issue 9433 - Deprecate delete

### PR DESCRIPTION
**Pull request by JinShil**
Spec and documentation update to accompany https://github.com/dlang/dmd/pull/7342

### CODE REVIEW

**Comment by MetaLang** [comment]
Even though `delete` has been "deprecated" for forever, I think this still needs approval from @andralex.

**Comment by JinShil** [comment]
This is ready for additional scrutiny and potential merge, now that https://github.com/dlang/druntime/pull/2037 is in.

**Comment by wilzbach** [comment]
FYI: When I deprecated the hex strings, I removed them from the spec immediately (https://github.com/dlang/dlang.org/pull/2118), which led to an outcry on the forums.
So while it's deprecated, we should probably just note that's deprecated and direct the users to destroy or __delete

**Comment by JinShil** [comment]
> So while it's deprecated, we should probably just note that's deprecated and direct the users to destroy or __delete

OK. all the documentation about `delete` in expression.dd remains, with the an added not to prefer `object.destory` or `__delete`, but other places in the documentation where `delete`'s use was advocated, I either removed it or replaced it with `object.destroy`.

File: spec/class.dd [hunk 1]
```diff
@@ -633,10 +633,10 @@ $(GNAME Destructor):
         )
 
         $(P The  program can explicitly inform the garbage collector that an
-        object is no longer referred to (with the delete expression), and
+        object is no longer referred to
+        (with $(LINK2 phobos/object.html#.destroy, $(D object.destroy()))), and
```
Comment (wilzbach) [inline spec/class.dd]: `$(REF1 destroy, object)
or if you like the parenthesis sth. like $(REF1_ALTTEXT, $(D object.destroy()), destroy, object)

File: spec/class.dd [hunk 2]
```diff
@@ -654,8 +654,9 @@ $(GNAME Destructor):
         members that are references to garbage collected objects, those
         references may no longer be valid. This means that destructors
         cannot reference sub objects.
-        This rule does not apply to auto objects or objects deleted
-        with the $(I DeleteExpression), as the destructor is not being run
+        This rule does not apply to auto objects or objects destructed
+        with $(LINK2 phobos/object.html#.destroy, $(D object.destroy())),
```
Comment (wilzbach) [inline spec/class.dd]: ditto

File: spec/expression.dd [hunk 3]
```diff
@@ -732,6 +732,9 @@ $(GRAMMAR
 $(GNAME DeleteExpression):
     $(D delete) $(GLINK UnaryExpression)
 )
+    $(P NOTE: `delete` has been deprecated.  Instead, please use $(REF1 destroy, object)
+    if feasible, or $(REF __delete, core, memory) as a last resort)
```
Comment (wilzbach) [inline spec/expression.dd]: End with `.`?

**Comment by JinShil** [comment]
Anyone know what's causing this error in DAutoTest?

```
LaTeX Warning: Hyper reference `delete_expressions' on page 163 undefined on in
put line 7533.

! You can't use `macro parameter character #' in horizontal mode.
l.7540 ... please use <a href="phobos/object.html#
                                                  .destroy">\lstinline|objec...

? 
! Emergency stop.
l.7540 ... please use <a href="phobos/object.html#
                                                  .destroy">\lstinline|objec...

!  ==> Fatal error occurred, no output PDF file produced!
```

File: spec/class.dd [hunk 4]
```diff
@@ -632,11 +632,10 @@ $(GNAME Destructor):
...
 
-        $(P The  program can explicitly inform the garbage collector that an
-        object is no longer referred to (with the delete expression), and
-        then the garbage collector calls the destructor
-        immediately, and adds the object's memory to the free storage.
-        The destructor is guaranteed to never be called twice.
+        $(P The program can explicitly inform the garbage collector that an
+        object is no longer referred to with $(REF1 destroy, object), and
+        then the garbage collector calls the destructor immediately.  The
+        destructor is guaranteed to never be called twice.
```
Comment (wilzbach) [inline spec/class.dd]: Missing parenthesis here.

**Comment by wilzbach** [comment]
> Anyone know what's causing this error in DAutoTest?

Yeah, it's the universally beloved DDoc system - it throws no error on trailing parenthesis, but LaTeX will.

File: spec/class.dd (hunk 4 above)
Comment (JinShil) [reply spec/class.dd]: The parens is there on the next line: 639.

**Comment by wilzbach** [comment]
Ah that's the real issue:

```
l.7540 ... please use <a href="phobos/object.html#
                                                  .destroy">\lstinline|objec...
```

Fix: https://github.com/dlang/dlang.org/pull/2194

**Comment by JinShil** [comment]
Thanks to @wilzbach this is now passing and is ready for additional scrutiny, and potentially a merge.

**Comment by JinShil** [comment]
Ping.  The DMD PR has been merged, and this really needs to go in along with it for the next release.

**Comment by JinShil** [comment]
@andralex or whoever can act on this.  This needs to go in with the other, already merged, PRs.

**Comment by andralex** [comment]
Cool, thanks.

**Pull request merged**